// @ibm-cloud/cloudant/auth/index.js
"use strict";
/**
 * © Copyright IBM Corporation 2020. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SessionTokenManager = exports.CouchdbSessionAuthenticator = exports.getAuthenticatorFromEnvironment = exports.IamTokenManager = exports.IamAuthenticator = exports.BasicAuthenticator = exports.NoAuthAuthenticator = void 0;
var ibm_cloud_sdk_core_1 = require("ibm-cloud-sdk-core");
Object.defineProperty(exports, "NoAuthAuthenticator", { enumerable: true, get: function () { return ibm_cloud_sdk_core_1.NoAuthAuthenticator; } });
Object.defineProperty(exports, "BasicAuthenticator", { enumerable: true, get: function () { return ibm_cloud_sdk_core_1.BasicAuthenticator; } });
Object.defineProperty(exports, "IamAuthenticator", { enumerable: true, get: function () { return ibm_cloud_sdk_core_1.IamAuthenticator; } });
Object.defineProperty(exports, "IamTokenManager", { enumerable: true, get: function () { return ibm_cloud_sdk_core_1.IamTokenManager; } });
Object.defineProperty(exports, "getAuthenticatorFromEnvironment", { enumerable: true, get: function () { return ibm_cloud_sdk_core_1.getAuthenticatorFromEnvironment; } });
var couchdbSessionAuthenticator_1 = require("./couchdbSessionAuthenticator");
Object.defineProperty(exports, "CouchdbSessionAuthenticator", { enumerable: true, get: function () { return couchdbSessionAuthenticator_1.CouchdbSessionAuthenticator; } });
var sessionTokenManager_1 = require("./sessionTokenManager");
Object.defineProperty(exports, "SessionTokenManager", { enumerable: true, get: function () { return sessionTokenManager_1.SessionTokenManager; } });
